core/src/main/java/com/linecorp/armeria/server/annotation/JavaBeanRequestConverterFunction.java

```diff
@@ -0,0 +1,241 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.annotation;
+
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.convertParameter;
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.httpParametersOf;
+import static com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.validateAndNormalizeSupportedType;
+import static org.reflections.ReflectionUtils.getAllFields;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import com.linecorp.armeria.common.AggregatedHttpMessage;
+import com.linecorp.armeria.common.HttpParameters;
+import com.linecorp.armeria.common.util.Holder;
+import com.linecorp.armeria.server.AnnotatedHttpServiceParamUtil.EnumConverter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.AsciiString;
+
+/**
+ * A default implementation of a {@link RequestConverterFunction} which converts the following parameters
+ * of the {@link AggregatedHttpMessage} to an object.
+ *
+ * <p>Path parameters to fields annotated by {@link Param}
+ *
+ * <p>HTTP parameters to fields annotated by {@link Param}
+ *
+ * <p>Headers to fields annotated by {@link Header}
+ */
+public class JavaBeanRequestConverterFunction implements RequestConverterFunction {
+    private static final ConcurrentMap<Class<?>, BeanParamInfo> beanParamInfoCache = new ConcurrentHashMap<>();
+
+    private static final List<Class<?>> ignoredResultTypeList = new ArrayList<>();
+
+    /**
+     * Initialization process for a expectedResultType.
+     *
+     * @param expectedResultType a candidate data type, which will converted by JavaBeanRequestConverterFunction
+     */
+    public static synchronized void initFor(final Class<?> expectedResultType) {
+        if (ignoredResultTypeList.contains(expectedResultType) ||
+            beanParamInfoCache.containsKey(expectedResultType)) {
+            // already processed this result type
+            return;
+        }
+
+        final BeanParamInfo beanParamInfo = buildBeanParamInfo(expectedResultType);
+        if (beanParamInfo == null) {
+            ignoredResultTypeList.add(expectedResultType);
+        } else {
+            beanParamInfoCache.put(expectedResultType, beanParamInfo);
+        }
+    }
+
+    private static BeanParamInfo buildBeanParamInfo(final Class<?> expectedResultType) {
+        final List<FieldParamInfo> fieldParamList = new ArrayList<>();
+        getAllFields(expectedResultType).forEach(
+                field -> {
+                    final Param param = field.getAnnotation(Param.class);
```

Comment:
Do we need to call `field.setAccessible(true)`?

Reply:
Yes, it is done in constructor of class `FieldParamInfo`